File: binance4j-connectors/src/main/java/com/binance4j/connectors/WebsocketConnectors.java
package com.binance4j.connectors;

import java.util.HashMap;
import java.util.Map;
import java.util.Optional;

import com.binance4j.core.dto.CandlestickInterval;
import com.binance4j.core.exception.ApiException;
import com.binance4j.websocket.callback.WebsocketCallback;
import com.binance4j.websocket.client.UserDataClient;
import com.binance4j.websocket.client.WebsocketAggTradeClient;
import com.binance4j.websocket.client.WebsocketAllBookTickersClient;
import com.binance4j.websocket.client.WebsocketAllMiniTickersClient;
import com.binance4j.websocket.client.WebsocketAllTickersClient;
import com.binance4j.websocket.client.WebsocketBookTickerClient;
import com.binance4j.websocket.client.WebsocketCandlestickClient;
import com.binance4j.websocket.client.WebsocketDepthClient;
import com.binance4j.websocket.client.WebsocketMiniDepthClient;
import com.binance4j.websocket.client.WebsocketMiniTickerClient;
import com.binance4j.websocket.client.WebsocketTradeClient;
import com.binance4j.websocket.client.WebsocketUserDataClient;
import com.binance4j.websocket.dto.AggTrade;
import com.binance4j.websocket.dto.BookTicker;
import com.binance4j.websocket.dto.Candle;
import com.binance4j.websocket.dto.Depth;
import com.binance4j.websocket.dto.DepthLevel;
import com.binance4j.websocket.dto.DepthUpdateSpeed;
import com.binance4j.websocket.dto.MiniDepth;
import com.binance4j.websocket.dto.MiniTicker;
import com.binance4j.websocket.dto.Ticker;
import com.binance4j.websocket.dto.Trade;
import com.binance4j.websocket.dto.UserDataUpdate;

/** Websocket connectors container. */
public class WebsocketConnectors {
	/** Inner {@link WebsocketAllBookTickersClient}. */
	private WebsocketAllBookTickersClient allBookTickers;
	/** Inner {@link WebsocketAllMiniTickersClient}. */
	private WebsocketAllMiniTickersClient allMiniTickers;
	/** Inner {@link WebsocketAllTickersClient}. */
	private WebsocketAllTickersClient allTickers;
	/** Inner {@link WebsocketAggTradeClient WebsocketAggTradeClients}. */
	private Map<String, WebsocketAggTradeClient> websocketAggTradeClients = new HashMap<>();
	/** Inner {@link WebsocketBookTickerClient WebsocketBookTickerClients}. */
	private Map<String, WebsocketBookTickerClient> websocketBookTickerClients = new HashMap<>();
	/** Inner {@link WebsocketCandlestickClient WebsocketCandlestickClients}. */
	private Map<String, WebsocketCandlestickClient> websocketCandlestickClients = new HashMap<>();
	/** Inner {@link WebsocketDepthClient WebsocketDepthClients}. */
	private Map<String, WebsocketDepthClient> websocketDepthClients = new HashMap<>();
	/** Inner {@link WebsocketMiniDepthClient WebsocketMiniDepthClients}. */
	private Map<String, WebsocketMiniDepthClient> websocketMiniDepthClients = new HashMap<>();
	/** Inner {@link WebsocketMiniTickerClient WebsocketMiniTickerClients}. */
	private Map<String, WebsocketMiniTickerClient> websocketMiniTickerClients = new HashMap<>();
	/** Inner {@link WebsocketTradeClient WebsocketTradeClients}. */
	private Map<String, WebsocketTradeClient> websocketTradeClients = new HashMap<>();
	/** Inner {@link WebsocketUserDataClient WebsocketUserDataClients}. */
	private Map<String, WebsocketUserDataClient> websocketUserDataClients = new HashMap<>();

	// allBookTickers //

	/**
	 * Creates an instance of {@link WebsocketAllBookTickersClient} and caches it.
	 * 
	 * @param callback Events handler.
	 * @return The generated client.
	 */
	public WebsocketAllBookTickersClient allBookTickers(WebsocketCallback<BookTicker> callback) {
		allBookTickers = new WebsocketAllBookTickersClient(callback);
		return allBookTickers;
	};

	/**
	 * @return The cached client.
	 */
	public Optional<WebsocketAllBookTickersClient> allBookTickers() {
		return Optional.of(allBookTickers);
	};

	// allMiniTickers //

	/**
	 * Creates an instance of {@link WebsocketAllMiniTickersClient} and caches it.
	 * 
	 * @param callback Events handler.
	 * @return The generated client.
	 */
	public WebsocketAllMiniTickersClient allMiniTickers(WebsocketCallback<MiniTicker> callback) {
		allMiniTickers = new WebsocketAllMiniTickersClient(callback);
		return allMiniTickers;
	};

	/**
	 * @return The cached client.
	 */
	public Optional<WebsocketAllMiniTickersClient> allMiniTickers() {
		return Optional.of(allMiniTickers);
	};

	// allTickers //

	/**
	 * Creates an instance of {@link WebsocketAllTickersClient} and caches it.
	 * 
	 * @param callback Events handler.
	 * @return The generated client.
	 */
	public WebsocketAllTickersClient allTickers(WebsocketCallback<Ticker> callback) {
		allTickers = new WebsocketAllTickersClient(callback);
		return allTickers;
	};

	/**
	 * @return The cached client.
	 */
	public Optional<WebsocketAllTickersClient> allTickers() {
		return Optional.of(allTickers);
	};

	// aggTrade //

	/**
	 * Creates an instance of {@link WebsocketAggTradeClient} and caches it.
	 * 
	 * @param symbols  Trading pairs to watch separated by a coma.
	 * @param callback Events handler.
	 * @return The generated client.
	 */
	public WebsocketAggTradeClient aggTrade(String symbols, WebsocketCallback<AggTrade> callback) {
		var client = new WebsocketAggTradeClient(symbols, callback);
		websocketAggTradeClients.put(client.getSymbols(), client);
		return client;
	};

	/**
	 * @param symbol The symbols the client is connected to.
	 * @return The cached client.
	 */
	public Optional<WebsocketAggTradeClient> aggTrade(String symbols) {
		return Optional.of(websocketAggTradeClients.get(handleSymbols(symbols)));
	};

	// bookTicker //

	/**
	 * Creates an instance of {@link WebsocketBookTickerClient} and caches it.
	 * 
	 * @param symbols  Trading pairs to watch separated by a coma.
	 * @param callback Events handler.
	 * @return The generated client.
	 */
	public WebsocketBookTickerClient bookTicker(String symbols, WebsocketCallback<BookTicker> callback) {
		var client = new WebsocketBookTickerClient(symbols, callback);
		websocketBookTickerClients.put(client.getSymbols(), client);
		return client;
	};

	/**
	 * @param symbol The symbols the client is connected to.
	 * @return The cached client.
	 */
	public Optional<WebsocketBookTickerClient> bookTicker(String symbols) {
		return Optional.of(websocketBookTickerClients.get(handleSymbols(symbols)));
	};

	// candlestick //

	/**
	 * Creates an instance of {@link WebsocketCandlestickClient} and caches it.
	 * 
	 * @param symbols  Trading pairs to watch separated by a coma.
	 * @param interval Candlestick interval.
	 * @param callback Events handler.
	 * @return The generated client.
	 */
	public WebsocketCandlestickClient candlestick(String symbols, CandlestickInterval interval,
			WebsocketCallback<Candle> callback) {
		var client = new WebsocketCandlestickClient(symbols, interval, callback);
		websocketCandlestickClients.put(String.format("%s%s", client.getSymbols(), interval), client);
		return client;
	};

	/**
	 * @param symbol The symbols the client is connected to.
	 * @return The cached client.
	 */
	public Optional<WebsocketCandlestickClient> candlestick(String symbols, CandlestickInterval interval) {
		return Optional.of(websocketCandlestickClients.get(String.format("%s%s", handleSymbols(symbols), interval)));
	};

	// depth //

	/**
	 * Creates an instance of {@link WebsocketDepthClient} and caches it.
	 * 
	 * @param symbols  Trading pairs to watch separated by a coma.
	 * @param speed    Update speed.
	 * @param callback Events handler.
	 * @return The generated client.
	 */
	public WebsocketDepthClient depth(String symbols, DepthUpdateSpeed speed, WebsocketCallback<Depth> callback) {
		var client = new WebsocketDepthClient(symbols, speed, callback);
		websocketDepthClients.put(String.format("%s%s", client.getSymbols(), speed), client);
		return client;
	};

	/**
	 * @param symbol The symbols the client is connected to.
	 * @return The cached client.
	 */
	public Optional<WebsocketDepthClient> depth(String symbols, DepthUpdateSpeed speed) {
		return Optional.of(websocketDepthClients.get(String.format("%s%s", handleSymbols(symbols), speed)));
	};

	// miniDepth //

	/**
	 * Creates an instance of {@link WebsocketMiniDepthClient} and caches it.
	 * 
	 * @param symbols  Trading pairs to watch separated by a coma.
	 * @param level    Order book size.
	 * @param speed    Update speed.
	 * @param callback Events handler.
	 * @return The generated client.
	 */
	public WebsocketMiniDepthClient miniDepth(String symbols, DepthLevel level, DepthUpdateSpeed speed,
			WebsocketCallback<MiniDepth> callback) {
		var client = new WebsocketMiniDepthClient(symbols, level, speed, callback);
		websocketMiniDepthClients.put(String.format("%s%s%s", client.getSymbols(), level, speed), client);
		return client;
	};

	/**
	 * @param symbol The symbols the client is connected to.
	 * @return The cached client.
	 */
	public Optional<WebsocketMiniDepthClient> miniDepth(String symbols, DepthLevel level, DepthUpdateSpeed speed) {
		return Optional
				.of(websocketMiniDepthClients.get(String.format("%s%s%s", handleSymbols(symbols), level, speed)));
	};

	// miniTicker //

	/**
	 * Creates an instance of {@link WebsocketMiniTickerClient} and caches it.
	 * 
	 * @param symbols  Trading pairs to watch separated by a coma.
	 * @param callback Events handler.
	 * @return The generated client.
	 */
	public WebsocketMiniTickerClient miniTicker(String symbols, WebsocketCallback<MiniTicker> callback) {
		var client = new WebsocketMiniTickerClient(symbols, callback);
		websocketMiniTickerClients.put(client.getSymbols(), client);
		return client;
	};

	/**
	 * @param symbol The symbols the client is connected to.
	 * @return The cached client.
	 */
	public Optional<WebsocketMiniTickerClient> miniTicker(String symbols) {
		return Optional.of(websocketMiniTickerClients.get(handleSymbols(symbols)));
	};

	// trade //

	/**
	 * Creates an instance of {@link WebsocketTradeClient} and caches it.
	 * 
	 * @param symbols  Trading pairs to watch separated by a coma.
	 * @param callback Events handler.
	 * @return The generated client.
	 */
	public WebsocketTradeClient trade(String symbols, WebsocketCallback<Trade> callback) {
		var client = new WebsocketTradeClient(symbols, callback);
		websocketTradeClients.put(client.getSymbols(), client);
		return client;
	};

	/**
	 * @param symbol The symbols the client is connected to.
	 * @return The cached client.
	 */
	public Optional<WebsocketTradeClient> trade(String symbols) {
		return Optional.of(websocketTradeClients.get(handleSymbols(symbols)));
	};

	// userData //

	/**
	 * Creates an instance of {@link WebsocketTradeClient} and caches it.
	 * 
	 * @param client   UserDataClient that will fetch the listen key to open the
	 *                 stream and keep it alive at a periodical interval.
	 * @param callback Events handler.
	 * @return The generated client.
	 */
	public WebsocketUserDataClient user(UserDataClient userDataClient, WebsocketCallback<UserDataUpdate> callback)
			throws ApiException {
		var client = new WebsocketUserDataClient(userDataClient, callback);
		websocketUserDataClients.put(client.getListenKey(), client);
		return client;
	};

	/**
	 * @param listenKey The listen key the client is connected to.
	 * @return The cached client.
	 */
	public Optional<WebsocketUserDataClient> user(String listenKey) {
		return Optional.of(websocketUserDataClients.get(listenKey));
	};

	private String handleSymbols(String symbols) {
		return symbols.replaceAll(" ", "").toLowerCase();
	}
}

File: src/main/kotlin/com/binance4j/connectors/WebsocketConnectors.kt
/*
 * MIT License
 *
 * Copyright (c) 2022 Binance4j
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NON INFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

package com.binance4j.connectors.connectors

import com.binance4j.connectors.core.dto.CandlestickInterval
import com.binance4j.connectors.core.exception.ApiException
import com.binance4j.connectors.websocket.callback.WebsocketCallback
import com.binance4j.connectors.websocket.client.*
import com.binance4j.connectors.websocket.dto.*
import java.util.*

/** Websocket connectors container.  */
object WebsocketConnectors {
	/** Inner [WebsocketAllBookTickersClient].  */
	private lateinit var allBookTickers: WebsocketAllBookTickersClient
	
	/** Inner [WebsocketAllMiniTickersClient].  */
	private lateinit var allMiniTickers: WebsocketAllMiniTickersClient
	
	/** Inner [WebsocketAllTickersClient].  */
	private  lateinit var allTickers: WebsocketAllTickersClient
	
	/** Inner [WebsocketAggTradeClients][WebsocketAggTradeClient].  */
	private val websocketAggTradeClients: MutableMap<String, WebsocketAggTradeClient> = HashMap()
	
	/** Inner [WebsocketBookTickerClients][WebsocketBookTickerClient].  */
	private val websocketBookTickerClients: MutableMap<String, WebsocketBookTickerClient> = HashMap()
	
	/** Inner [WebsocketCandlestickClients][WebsocketCandlestickClient].  */
	private val websocketCandlestickClients: MutableMap<String, WebsocketCandlestickClient> = HashMap()
	
	/** Inner [WebsocketDepthClients][WebsocketDepthClient].  */
	private val websocketDepthClients: MutableMap<String, WebsocketDepthClient> = HashMap()
	
	/** Inner [WebsocketMiniDepthClients][WebsocketMiniDepthClient].  */
	private val websocketMiniDepthClients: MutableMap<String, WebsocketMiniDepthClient> = HashMap()
	
	/** Inner [WebsocketMiniTickerClients][WebsocketMiniTickerClient].  */
	private val websocketMiniTickerClients: MutableMap<String, WebsocketMiniTickerClient> = HashMap()
	
	/** Inner [WebsocketTradeClients][WebsocketTradeClient].  */
	private val websocketTradeClients: MutableMap<String, WebsocketTradeClient> = HashMap()
	
	/** Inner [WebsocketUserDataClients][WebsocketUserDataClient].  */
	private val websocketUserDataClients: MutableMap<String, WebsocketUserDataClient> = HashMap()
	
	// allBookTickers //
	
	/**
	 * Creates an instance of [WebsocketAllBookTickersClient] and caches it.
	 *
	 * @param callback Events handler.
	 * @return The generated client.
	 */
	fun allBookTickers(callback: WebsocketCallback<BookTicker>): WebsocketAllBookTickersClient {
		allBookTickers = WebsocketAllBookTickersClient(callback)
		return allBookTickers
	}
	
	/**
	 * @return The cached client.
	 */
	fun allBookTickers() =  allBookTickers

	// allMiniTickers //
	
	/**
	 * Creates an instance of [WebsocketAllMiniTickersClient] and caches it.
	 *
	 * @param callback Events handler.
	 * @return The generated client.
	 */
	fun allMiniTickers(callback: WebsocketCallback<MiniTicker>): WebsocketAllMiniTickersClient {
		allMiniTickers = WebsocketAllMiniTickersClient(callback)
		return allMiniTickers
	}
	
	/**
	 * @return The cached client.
	 */
	fun allMiniTickers(): Optional<WebsocketAllMiniTickersClient> {
		return Optional.of(allMiniTickers)
	}
	
	// allTickers //
	
	/**
	 * Creates an instance of [WebsocketAllTickersClient] and caches it.
	 *
	 * @param callback Events handler.
	 * @return The generated client.
	 */
	fun allTickers(callback: WebsocketCallback<Ticker>): WebsocketAllTickersClient {
		allTickers = WebsocketAllTickersClient(callback)
		return allTickers
	}
	
	/**
	 * @return The cached client.
	 */
	fun allTickers(): WebsocketAllTickersClient = allTickers

	// aggTrade //
	
	/**
	 * Creates an instance of [WebsocketAggTradeClient] and caches it.
	 *
	 * @param symbols  Trading pairs to watch separated by a coma.
	 * @param callback Events handler.
	 * @return The generated client.
	 */
	fun aggTrade(symbols: String, callback: WebsocketCallback<AggTrade>): WebsocketAggTradeClient {
		val client = WebsocketAggTradeClient(symbols, callback)
		websocketAggTradeClients[client.symbols!!] = client
		return client
	}
	
	/**
	 * @param symbols The symbols the client is connected to.
	 * @return The cached client.
	 */
	fun aggTrade(symbols: String): WebsocketAggTradeClient? {
		return websocketAggTradeClients[handleSymbols(symbols)]
	}

	// bookTicker //

	/**
	 * Creates an instance of [WebsocketBookTickerClient] and caches it.
	 *
	 * @param symbols  Trading pairs to watch separated by a coma.
	 * @param callback Events handler.
	 * @return The generated client.
	 */
	fun bookTicker(symbols: String, callback: WebsocketCallback<BookTicker>): WebsocketBookTickerClient {
		val client = WebsocketBookTickerClient(symbols, callback)
		websocketBookTickerClients[client.symbols!!] = client
		return client
	}

	/**
	 * @param symbols The symbols the client is connected to.
	 * @return The cached client.
	 */
	fun bookTicker(symbols: String): WebsocketBookTickerClient? {
		return websocketBookTickerClients[handleSymbols(symbols)]
	}

	// candlestick //

	/**
	 * Creates an instance of [WebsocketCandlestickClient] and caches it.
	 *
	 * @param symbols  Trading pairs to watch separated by a coma.
	 * @param interval Candlestick interval.
	 * @param callback Events handler.
	 * @return The generated client.
	 */
	fun candlestick(
		symbols: String,
		interval: CandlestickInterval,
		callback: WebsocketCallback<Candle>
	): WebsocketCandlestickClient {
		val client = WebsocketCandlestickClient(symbols, interval, callback)
		websocketCandlestickClients[String.format("%s%s", client.symbols, interval)] = client
		return client
	}

	/**
	 * @param symbols The symbols the client is connected to.
	 * @return The cached client.
	 */
	fun candlestick(symbols: String, interval: CandlestickInterval): WebsocketCandlestickClient? {
		return websocketCandlestickClients[String.format("%s%s", handleSymbols(symbols), interval)]
	}

	// depth //

	/**
	 * Creates an instance of [WebsocketDepthClient] and caches it.
	 *
	 * @param symbols  Trading pairs to watch separated by a coma.
	 * @param speed    Update speed.
	 * @param callback Events handler.
	 * @return The generated client.
	 */
	fun depth(symbols: String, speed: DepthUpdateSpeed, callback: WebsocketCallback<Depth>): WebsocketDepthClient {
		val client = WebsocketDepthClient(symbols, speed, callback)
		websocketDepthClients[String.format("%s%s", client.symbols, speed)] = client
		return client
	}

	/**
	 * @param symbols The symbols the client is connected to.
	 * @return The cached client.
	 */
	fun depth(symbols: String, speed: DepthUpdateSpeed): WebsocketDepthClient? {
		return websocketDepthClients[String.format("%s%s", handleSymbols(symbols), speed)]
	}

	// miniDepth //

	/**
	 * Creates an instance of [WebsocketMiniDepthClient] and caches it.
	 *
	 * @param symbols  Trading pairs to watch separated by a coma.
	 * @param level    Order book size.
	 * @param speed    Update speed.
	 * @param callback Events handler.
	 * @return The generated client.
	 */
	fun miniDepth(
		symbols: String, level: DepthLevel, speed: DepthUpdateSpeed,
		callback: WebsocketCallback<MiniDepth>
	): WebsocketMiniDepthClient {
		val client = WebsocketMiniDepthClient(symbols, level, speed, callback)
		websocketMiniDepthClients[String.format("%s%s%s", client.symbols, level, speed)] = client
		return client
	}

	/**
	 * @param symbols The symbols the client is connected to.
	 * @return The cached client.
	 */
	fun miniDepth(symbols: String, level: DepthLevel, speed: DepthUpdateSpeed): WebsocketMiniDepthClient? {
		return websocketMiniDepthClients[String.format("%s%s%s", handleSymbols(symbols), level, speed)]
	}

	// miniTicker //

	/**
	 * Creates an instance of [WebsocketMiniTickerClient] and caches it.
	 *
	 * @param symbols  Trading pairs to watch separated by a coma.
	 * @param callback Events handler.
	 * @return The generated client.
	 */
	fun miniTicker(symbols: String, callback: WebsocketCallback<MiniTicker>): WebsocketMiniTickerClient {
		val client = WebsocketMiniTickerClient(symbols, callback)
		websocketMiniTickerClients[client.symbols!!] = client
		return client
	}

	/**
	 * @param symbols The symbols the client is connected to.
	 * @return The cached client.
	 */
	fun miniTicker(symbols: String): WebsocketMiniTickerClient? {
		return websocketMiniTickerClients[handleSymbols(symbols)]
	}

	// trade //

	/**
	 * Creates an instance of [WebsocketTradeClient] and caches it.
	 *
	 * @param symbols  Trading pairs to watch separated by a coma.
	 * @param callback Events handler.
	 * @return The generated client.
	 */
	fun trade(symbols: String, callback: WebsocketCallback<Trade>): WebsocketTradeClient {
		val client = WebsocketTradeClient(symbols, callback)
		websocketTradeClients[client.symbols!!] = client
		return client
	}

	/**
	 * @param symbols The symbols the client is connected to.
	 * @return The cached client.
	 */
	fun trade(symbols: String): WebsocketTradeClient? {
		return websocketTradeClients[handleSymbols(symbols)]
	}

	// userData //

	/**
	 * Creates an instance of [WebsocketTradeClient] and caches it.
	 *
	 * @param userDataClient UserDataClient that will fetch the listen key to open
	 * the
	 * stream and keep it alive at a periodical interval.
	 * @param callback       Events handler.
	 * @return The generated client.
	 */
	@Throws(ApiException::class)
	fun user(userDataClient: UserDataClient, callback: WebsocketCallback<UserDataUpdate>): WebsocketUserDataClient {
		val client = WebsocketUserDataClient(userDataClient, callback)
		websocketUserDataClients[client.listenKey] = client
		return client
	}

	/**
	 * @param listenKey The listen key the client is connected to.
	 * @return The cached client.
	 */
	fun user(listenKey: String): WebsocketUserDataClient? {
		return websocketUserDataClients[listenKey]
	}
	
	private fun handleSymbols(symbols: String): String {
		return symbols.replace(" ", "").lowercase(Locale.getDefault())
	}
}